### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by Charana123**
instructions - 

    download packages - npm install
    install grunt globally (only done one) - npm install grunt
    transpile javascript - grunt
    run tests - npm test

### CODE REVIEW

File: spec/monasca_client_tests.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import nock from 'nock';
+
+export class monasca_api_mock {
```
Comment (dougszumski) [inline spec/monasca_client_tests.js]: Is this left over from testing? It doesn't look like we use it? Can we delete it?

File: src/components/alarm_definitions.js [hunk 2]
```diff
@@ -22,9 +22,9 @@ import MonascaClient from './monasca_client';
 export class AlarmDefinitionsPageCtrl {
 
   /** @ngInject */
-  constructor($scope, $injector, backendSrv, datasourceSrv, alertSrv) {
+  constructor($scope, $injector, backendSrv, datasourceSrv, alertSrv, monascaClientSrv) {
```
Comment (dougszumski) [inline src/components/alarm_definitions.js]: Can we avoid modifying the code that we're testing?

File: src/components/edit_alarm_definition.js [hunk 3]
```diff
@@ -73,13 +73,13 @@ export class EditAlarmDefinitionPageCtrl {
 
     this.monasca.getAlarmDefinition(this.id)
       .then(alarm_definition => {
-	this.savedAlarmDefinition = this.pickKnownFields(alarm_definition);
-	this.newAlarmDefinition = _.cloneDeep(this.savedAlarmDefinition);
+	      this.savedAlarmDefinition = this.pickKnownFields(alarm_definition);
```
Comment (dougszumski) [inline src/components/edit_alarm_definition.js]: Please put all re-formatting commits in a separate commit. The idea is that it makes the actual changes to the code easier to pick out.

File: src/components/monasca_client.js [hunk 4]
```diff
@@ -235,59 +234,62 @@ export default class MonascaClient {
...
-	}
-	return this.datasourceSrv.get(response.jsonData.datasourceName)
-	  .then(ds => {
-	    this.ds = ds;
-	    return this.ds;
-	  })
-	  .catch(err => { throw err; });
-      })
-      .catch(err => { throw err; });
+    .then(response => {
```
Comment (dougszumski) [inline src/components/monasca_client.js]: Please put this reformatting change in a separate commit.

**Review by dougszumski — CHANGES_REQUESTED** [review]
Thanks @Charana123! A couple of nits to fix before we get started..

File: src/components/alarm_definitions.js (hunk 2 above)
Comment (Charana123) [reply src/components/alarm_definitions.js]: you can't test the grafana components without dependency injection in the constructor

File: src/components/config.js [hunk 5]
```diff
@@ -21,7 +21,7 @@ export class MonascaAppConfigCtrl {
       this.appModel.jsonData = {};
     }
 
-    var show_types = ['monasca-grafana-datasource', 'monasca-datasource'];
+    var show_types = ['monasca-grafana-datasource', 'monasca-datasource', 'test-'];
```
Comment (dougszumski) [inline src/components/config.js]: I don't think this should be in the commit?

File: src/components/edit_alarm_definition.js [hunk 6]
```diff
@@ -22,10 +22,10 @@ import _ from 'lodash';
 export class EditAlarmDefinitionPageCtrl {
 
   /** @ngInject */
-  constructor($scope, $injector, $location, backendSrv, datasourceSrv, alertSrv) {
+  constructor($scope, $injector, $location, backendSrv, datasourceSrv, alertSrv, monascaClientSrv) {
```
Comment (dougszumski) [inline src/components/edit_alarm_definition.js]: Can we remove backendSrv, datasourceSrv now that you're injecting the monasca client? They don't appear to be used?

File: src/components/edit_notification.js [hunk 7]
```diff
@@ -21,10 +21,10 @@ import MonascaClient from './monasca_client';
 export class EditNotificationPageCtrl {
 
   /** @ngInject */
-  constructor($scope, $injector, $location, backendSrv, datasourceSrv, alertSrv) {
+  constructor($scope, $injector, $location, backendSrv, datasourceSrv, alertSrv, monascaClientSrv) {
```
Comment (dougszumski) [inline src/components/edit_notification.js]: Can we remove backendSrv, datasourceSrv now that you're injecting the monasca client? They don't appear to be used?
(Same for other instances)

File: src/components/monasca_client.js [hunk 8]
```diff
@@ -235,59 +234,62 @@ export default class MonascaClient {
...
-          if (monasca_response) {
-            throw { message: 'Monasca Error Response: ' + monasca_response };
-          } else {
-            throw { message: 'Monasca Error Status: ' + err.status };
-          }
-	}
-      });
-    }).catch(err => { throw err; });
+    return Promise.reject( { message: 'MY ERROR!'} )
+  //   return this._getDataSource().then(data_source => {
```
Comment (dougszumski) [inline src/components/monasca_client.js]: I guess commenting this out isn't supposed to be in the commit?

If you haven't figured it out already (I'm sure you have :) ) you can use ```git rebase -i <commit_id>``` to go back and amend previous commits.

File: src/components/edit_alarm_definition.js (hunk 6 above)
Comment (brtknr) [reply src/components/edit_alarm_definition.js]: +1

File: src/components/monasca_client.js [hunk 9]
```diff
@@ -15,8 +15,8 @@
  *   limitations under the License.
  */
 
-import config from 'app/core/config';
-import appEvents from 'app/core/app_events';
+// import config from 'app/core/config';
```
Comment (brtknr) [inline src/components/monasca_client.js]: Remove these if not needed

File: src/components/monasca_client.js [hunk 10]
```diff
@@ -15,8 +15,8 @@
...
+// import appEvents from 'app/core/app_events';
```
Comment (brtknr) [inline src/components/monasca_client.js]: This too

File: src/spec/monasca_client_spec.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+// }
+
+
+function backendSrvMock(){
+
+}
+
+function datasourceSrvMock(){
+
+}
```
Comment (brtknr) [inline src/spec/monasca_client_spec.js]: Will backendSrvMock and datasourceSrvMock be necessary after removing them as inputs?

**Comment by brtknr** [comment]
> download packages - npm install
> install grunt globally (only done one) - npm install grunt
> transpile javascript - grunt
> run tests - npm test

Add these to README

**Comment by matrixik** [comment]
Could you use `Start a review` instead of `Add single comment`?
This way github won't be sending so many mails...

File: src/components/monasca_client.js [hunk 12]
```diff
@@ -20,14 +20,14 @@ import appEvents from 'app/core/app_events';
 
 export default class MonascaClient {
 
+  /** @ngInject */
```
Comment (dougszumski) [inline src/components/monasca_client.js]: Please move this out of the reformat commit.

File: src/components/monasca_client.js (hunk 4 above)
Comment (dougszumski) [reply src/components/monasca_client.js]: This still needs to be moved to the reformat commit.

File: src/components/monasca_client.js [hunk 13]
```diff
@@ -233,59 +228,60 @@ export default class MonascaClient {
...
   }
 
   _request(method, path, params, data) {
     return this._getDataSource().then(data_source => {
 
       var headers = {
-	'Content-Type': 'application/json',
-	'X-Auth-Token': data_source.token
+	      'Content-Type': 'application/json',
+	      'X-Auth-Token': data_source.token
```
Comment (dougszumski) [inline src/components/monasca_client.js]: Should be in the reformat commit.

File: src/components/monasca_client.js [hunk 14]
```diff
@@ -233,59 +228,60 @@ export default class MonascaClient {
...
           var monasca_response;
           if (err.data) {
             if (err.data.message){
               monasca_response = err.data.message;
-	    }
-	    else if (err.data.description) {
-	      monasca_response = err.data.description;
-	    } else{
+	          }
+	          else if (err.data.description) {
```
Comment (dougszumski) [inline src/components/monasca_client.js]: Same for this..

File: Gruntfile.js [hunk 15]
```diff
@@ -64,3 +64,6 @@ module.exports = function(grunt) {
 
   grunt.registerTask('default', ['clean', 'sass', 'copy', 'babel']);
 };
+
```
Comment (dougszumski) [inline Gruntfile.js]: Please remove - it's good practice to keep change requests as small as possible.

File: karma.conf.js [hunk 16]
```diff
@@ -69,8 +73,10 @@ module.exports = function(config) {
 
     // start these browsers
     // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
-    browsers: ['Chrome'],
-
+    browsers: [
+      // 'Chrome'
```
Comment (dougszumski) [inline karma.conf.js]: I think we can delete this line.

File: src/spec/monasca_client.spec.js [hunk 17]
```diff
@@ -0,0 +1,14 @@
+import {MonascaClient} from "../src/components/monasca_client"
+
+describe("Monasca Unit Tests", () => {
+
+    it("test", () => {
```
Comment (dougszumski) [inline src/spec/monasca_client.spec.js]: Looks like this is temporary code which can be removed now.

File: karma-test-shim.js [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+jasmine.DEFAULT_TIMEOUT_INTERVAL = 1000;
+
+// (1)
+var builtPaths = (__karma__.config.builtPaths || ['app/'])
+                  .map(function(p) { return '/base/'+p;});
+
+// (2)
+__karma__.loaded = function () { };
+
+// (3)
```
Comment (dougszumski) [inline karma-test-shim.js]: In general, comments should make sense to anyone reading the code. I'm not sure why these are numbered - please either explain why, or remove the comments.

File: karma.conf.js [hunk 19]
```diff
@@ -10,20 +10,33 @@ module.exports = function(config) {
 
     // frameworks to use
     // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
-    frameworks: ['jasmine'],
+    frameworks: [ 
+      'jasmine'
```
Comment (dougszumski) [inline karma.conf.js]: Please move this to the reformatting commit.

File: src/spec/monasca_client.spec.js [hunk 20]
```diff
@@ -1,14 +0,0 @@
-import {MonascaClient} from "../src/components/monasca_client"
-
-describe("Monasca Unit Tests", () => {
```
Comment (dougszumski) [inline src/spec/monasca_client.spec.js]: It looks like this belongs in the previous commit where it was added (since we want to remove it)

File: src/spec/monasca_client_spec.js [hunk 21]
```diff
@@ -0,0 +1,12 @@
+import {MonascaClient} from "../components/monasca_client.js"
+
+describe('Something abstract', function() {
+    describe('Something specific', function() {
```
Comment (dougszumski) [inline src/spec/monasca_client_spec.js]: This looks like temporary test code? We don't want to merge it into master.

File: src/components/monasca_client.js [hunk 22]
```diff
@@ -19,207 +19,148 @@ export default class MonascaClient {
...
-        if(dimensions[i].severity){
-          if(params.severity == ""){
-            params.severity = dimensions[i].severity;
-          }
-          else{
-            params.severity += "|" + dimensions[i].severity;
-          }
-
-        }
+    params.metric_dimensions = query_parameters.metric_dimensions ? query_parameters.metric_dimensions.join(",") : undefined;
```
Comment (dougszumski) [inline src/components/monasca_client.js]: This is much neater than the original code! Good job.

File: src/components/monasca_client.js [hunk 23]
```diff
@@ -19,207 +19,148 @@ export default class MonascaClient {
...
-    return this._delete('/v2.0/alarm-definitions/' + id)
-      .then(resp => null)
-      .catch(err => { throw err; });
+    return (id == undefined ? Promise.reject("no id given to alarm definition patch request") : Promise.resolve(id))
+      .then(() => this._delete('/v2.0/alarm-definitions/' + id))
+      .then(resp => undefined)
   }
 
   // Notification Method Types
-
```
Comment (dougszumski) [inline src/components/monasca_client.js]: More reformatting..needs to be moved to reformat commit.

File: src/components/monasca_client.js [hunk 24]
```diff
@@ -19,207 +19,148 @@ export default class MonascaClient {
...
-      .then(resp => resp.data)
-      .catch(err => { throw err; });
+    return (id == undefined ? Promise.reject("No id given to alarm resource get request") : Promise.resolve(id))
+    .then(() => this._get('/v2.0/alarms/' + id))
+    .then(resp => resp.data)
   }
 
   getAlarmHistory(id){
-    return this._get('/v2.0/alarms/' + id + '/state-history')
+    return (id == undefined ? Promise.reject("no id given to alarm history get request") : Promise.resolve(id))
```
Comment (dougszumski) [inline src/components/monasca_client.js]: Good stuff - validating the id before making the call is a good improvement.

File: src/components/monasca_client.js [hunk 25]
```diff
@@ -19,207 +19,148 @@ export default class MonascaClient {
...
   };
 
   _post(path, data) {
-    return this._request('POST', path, null, data);
+    return this._request('POST', path, undefined, data);
   };
 
   _patch(path, data) {
-    return this._request('PATCH', path, null, data);
+    return this._request('PATCH', path, undefined, data);
```
Comment (dougszumski) [inline src/components/monasca_client.js]: Please explain why we're changing this from null.

File: src/components/monasca_client.js [hunk 26]
```diff
@@ -271,18 +212,19 @@ export default class MonascaClient {
...
+              }
             }
           }
-        if (monasca_response) {
-          throw { message: 'Monasca Error Response: ' + monasca_response };
-        } 
-        else {
-          throw { message: 'Monasca Error Status: ' + err.status };
+          if (monasca_response) {
+            throw { message: 'Monasca Error Response: ' + monasca_response };
```
Comment (dougszumski) [inline src/components/monasca_client.js]: This needs to go to the reformatting commit..

**Review by dougszumski — CHANGES_REQUESTED** [review]
Great work adding the tests Charana - it's especially nice to see them running in Travis. After you've addressed the comments I'll go through them. Thanks.

File: src/spec/monasca_client/notification_methods_spec.js [hunk 27]
```diff
@@ -0,0 +1,344 @@
...
+                    jsonData: {
+                        datasourceName: "monasca_datasource"
+                    }
+                }))
+                backendsrv_mock.datasourceRequest.and.returnValue(Promise.resolve({
+                    data: {
+                        elements: [
+                            { type: "EMAIL" },
+                            { type: "PAGERDUTY" },
+                            { type: "WEBHOOK"}
```
Comment (dougszumski) [inline src/spec/monasca_client/notification_methods_spec.js]: We can add Slack etc in a different commit..

**Review by dougszumski — APPROVED** [review]
I think this is looking pretty good in general. It would be nice to squash the commits into three commits. One for adding the tests, travis and the framework, the second for reformatting commits and the third for the changes to the source code. Don't worry if it's going to cause you a lot of conflicts.

File: src/components/alarms.js [hunk 28]
```diff
@@ -100,9 +95,9 @@ export class AlarmsPageCtrl {
     this.metricFilters.push({});
   }
 
-  removeMetricFilter(index) {
-    var filter = this.metricFilters[index];
-    this.metricFilters.splice(index, 1);
+  removeMetricFilter (index) {
+    // var filter = this.metricFilters[index]
```
Comment (dougszumski) [inline src/components/alarms.js]: nit: We should remove this if we don't need it

File: src/components/alarms.js [hunk 29]
```diff
@@ -122,9 +117,9 @@ export class AlarmsPageCtrl {
     this.severityFilters.push({});
   }
 
-  removeSeverityFilter(index) {
-    var filter = this.severityFilters[index];
-    this.severityFilters.splice(index, 1);
+  removeSeverityFilter (index) {
+    // var filter = this.severityFilters[index]
```
Comment (dougszumski) [inline src/components/alarms.js]: nit: We should remove this if we don't need it

File: src/components/alarms.js [hunk 30]
```diff
@@ -111,9 +106,9 @@ export class AlarmsPageCtrl {
     this.stateFilters.push({});
   }
 
-  removeStateFilter(index) {
-    var filter = this.stateFilters[index];
-    this.stateFilters.splice(index, 1);
+  removeStateFilter (index) {
+    // var filter = this.stateFilters[index]
```
Comment (dougszumski) [inline src/components/alarms.js]: nit: We should remove this if we don't need it

**Review by dougszumski — COMMENTED** [review]
Please can we squash the commits?

File: src/components/alarm_definitions.js (hunk 2 above)
Comment (dougszumski) [inline src/components/alarm_definitions.js]: nit: Since this change makes backendSrv, datasourceSrv redundant it would be better to remove them in this commit rather than the next one.

File: src/components/monasca_client.js [hunk 31]
```diff
@@ -101,60 +96,52 @@ export default class MonascaClient {
...
       .then(resp => resp.data.elements)
-      .catch(err => { throw err; });
     }
   }
 
   deleteAlarm(id) {
     return this._delete('/v2.0/alarms/' + id)
       .then(resp => null)
-      .catch(err => { throw err; });
+
```
Comment (dougszumski) [inline src/components/monasca_client.js]: nit: extra space
